pkg/mcs/resource_manager/server/server.go

```diff
@@ -0,0 +1,293 @@
+
+	serverLoopWg sync.WaitGroup
+	// Callback functions for different stages
+	// startCallbacks will be called after the server is started.
+	startCallbacks []func()
+	// primaryCallbacks will be called after the server becomes leader.
+	primaryCallbacks []func(context.Context)
+}
+
+// Name returns the unique etcd Name for this server in etcd cluster.
+func (s *Server) Name() string {
+	return s.name
+}
+
+// Context returns the context.
+func (s *Server) Context() context.Context {
+	return s.ctx
+}
+
+// Run runs the pd server.
+func (s *Server) Run() error {
+	if err := s.initClient(); err != nil {
+		return err
+	}
+	if err := s.startServer(); err != nil {
+		return err
+	}
+	return s.mux.Serve()
+}
+
+// Close closes the server.
+func (s *Server) Close() {
+	if !atomic.CompareAndSwapInt64(&s.isServing, 1, 0) {
+		// server is already closed
+		return
+	}
+
+	log.Info("closing server")
+
+	s.grpcServer.Stop()
+	s.httpServer.Shutdown(s.ctx)
+	s.serverLoopWg.Wait()
+
+	if s.etcdClient != nil {
+		if err := s.etcdClient.Close(); err != nil {
+			log.Error("close etcd client meet error", errs.ZapError(errs.ErrCloseEtcdClient, err))
+		}
+	}
+
+	if s.httpClient != nil {
+		s.httpClient.CloseIdleConnections()
+	}
+}
+
+// GetClient returns builtin etcd client.
+func (s *Server) GetClient() *clientv3.Client {
+	return s.etcdClient
+}
+
+// GetHTTPClient returns builtin http client.
+func (s *Server) GetHTTPClient() *http.Client {
+	return s.httpClient
+}
+
+// AddStartCallback adds a callback in the startServer phase.
+func (s *Server) AddStartCallback(callbacks ...func()) {
+	s.startCallbacks = append(s.startCallbacks, callbacks...)
+}
+
+// IsServing returns whether the server is the leader, if there is embedded etcd, or the primary otherwise.
+func (s *Server) IsServing() bool {
+	// TODO: implement this
+	return true
+}
+
+// AddServiceReadyCallback adds the callback function when the server becomes the leader, if there is embedded etcd, or the primary otherwise.
+func (s *Server) AddServiceReadyCallback(callbacks ...func(context.Context)) {
+	s.primaryCallbacks = append(s.primaryCallbacks, callbacks...)
+}
+
+func (s *Server) initClient() error {
+	tlsConfig, err := s.cfg.Security.ToTLSConfig()
+	if err != nil {
+		return err
+	}
+	endpoint, err := url.Parse(s.cfg.BackendEndpoints)
+	if err != nil {
+		return err
+	}
+	s.etcdClient, s.httpClient, err = etcdutil.CreateClients(tlsConfig, []url.URL{*endpoint})
+	return err
+}
+
+func (s *Server) startServer() error {
+	tlsConfig, err := s.cfg.Security.ToTLSConfig()
+	if err != nil {
+		return err
+	}
+	if tlsConfig != nil {
+		l, err := tls.Listen("tcp", s.cfg.ListenAddr, tlsConfig)
+		if err != nil {
+			return err
+		}
+		s.mux = cmux.New(l)
+	} else {
+		l, err := net.Listen("tcp", s.cfg.ListenAddr)
+		if err != nil {
+			return err
+		}
+		s.mux = cmux.New(l)
+	}
+
+	grpcL := s.mux.MatchWithWriters(cmux.HTTP2MatchHeaderFieldSendSettings("content-type", "application/grpc"))
```

Comment:
What is the difference between `HTTP2MatchHeaderFieldSendSettings` and `HTTP2`?

Reply:
It will block requests in tests. And In `[cmux](https://github.com/soheilhy/cmux)` doc, it also advice simiarly.

Reply:
etcd uses `HTTP2` instead.